File: E-Commerce App Project (Tabbed)/Classes/CategoryItemList.m
//
//  CategoryItemList.m
//  E-Commerce App Project (Tabbed)
//
//  Created by Rony Banik on 10/11/18.
//  Copyright © 2018 Rony Banik. All rights reserved.
//

#import "CategoryItemList.h"
#import "Item.h"
#import "Items.h"

@implementation CategoryItemList

+(CategoryItemList *)sharedInstance{
    static CategoryItemList* _sharedCategoryList;
    
    static dispatch_once_t once;
    
    dispatch_once(&once, ^{
        _sharedCategoryList = [[CategoryItemList alloc] init];
    });
    return _sharedCategoryList;
}

- (id)init
{
    if((self = [super init]))
    {
        NSArray *allItemsList = [Items sharedInstance].allItems;
        
        _tvCategoryItemsList = [self returnCategoryArray:allItemsList forCategory:@"Television"];
        _laptopCategoryItemsList= [self returnCategoryArray:allItemsList forCategory:@"Laptops"];
        _desktopCategoryItemsList= [self returnCategoryArray:allItemsList forCategory:@"Desktop"];
        _mobileCategoryItemsList= [self returnCategoryArray:allItemsList forCategory:@"Mobile"];
        _tabletCategoryItemsList= [self returnCategoryArray:allItemsList forCategory:@"Tablet"];
    }
    return self;
}


-(NSArray *)returnCategoryArray:(NSArray *)allItems
                    forCategory:(NSString *)itemCategoryName{
    
    NSMutableArray *theCategory;
    theCategory = [allItems mutableCopy];
    
    Item* runningItem;
    
    for (int i=0; i<allItems.count; i++) {
        if (![[allItems[i] itemCategory]  isEqual: itemCategoryName]) {
            runningItem=allItems[i];
            [theCategory removeObject:allItems[i]];
        }
    }
    return theCategory;
}
@end

File: E-Commerce App Project (Tabbed)/Classes/CategoryItemList.swift
//
//  CategoryItemList.swift
//  E-Commerce App Project (Tabbed)
//
//  Converted to Swift
//

import Foundation

@objc class CategoryItemList: NSObject {
    @objc static let sharedInstance = CategoryItemList()
    
    @objc private(set) var tvCategoryItemsList: [Item] = []
    @objc private(set) var laptopCategoryItemsList: [Item] = []
    @objc private(set) var mobileCategoryItemsList: [Item] = []
    @objc private(set) var desktopCategoryItemsList: [Item] = []
    @objc private(set) var tabletCategoryItemsList: [Item] = []
    
    private override init() {
        super.init()
        let allItemsList = Items.sharedInstance.allItems
        
        tvCategoryItemsList = returnCategoryArray(allItems: allItemsList, forCategory: "Television")
        laptopCategoryItemsList = returnCategoryArray(allItems: allItemsList, forCategory: "Laptops")
        desktopCategoryItemsList = returnCategoryArray(allItems: allItemsList, forCategory: "Desktop")
        mobileCategoryItemsList = returnCategoryArray(allItems: allItemsList, forCategory: "Mobile")
        tabletCategoryItemsList = returnCategoryArray(allItems: allItemsList, forCategory: "Tablet")
    }
    
    private func returnCategoryArray(allItems: [Item], forCategory itemCategoryName: String) -> [Item] {
        return allItems.filter { $0.itemCategory == itemCategoryName }
    }
}
